[te] Add a 2D convolution example test

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45906 [te][llvm] Enable fused multiply-add (fma) in code generation
* #45905 [te] Tiled (m=32 x n=32) gemm benchmark
* #45875 [te] Add a benchmark harness
* **#45514 [te] Add a 2D convolution example test**



Differential Revision: [D24142405](https://our.internmc.facebook.com/intern/diff/D24142405)

bertmaher committed:
[te] Add a 2D convolution example test

[ghstack-poisoned]

zheng-xq commented on test/cpp/tensorexpr/test_conv.cpp:
```diff
@@ -0,0 +1,88 @@
+        // because the `operator()` overload is only specialized for a small
+        // number of arguments.
+        return inputB.call(
+                   std::vector<te::ExprHandle>({n, c, oh + r, ow + s})) *
```
We need the handling of the paddings here. I am a bit surprised that this didn't OOB....

bertmaher replied:
Hm really?  This compute expression is an unpadded conv (or at least it's supposed to be).  Since `OH=H-R+1` the max extent of `oh + r` would be. (\me does algebra) `(H-R+1)-1 + (R-1)` or `H-1`, which is in bounds.  Right?

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_conv.cpp:
```diff
@@ -0,0 +1,88 @@
+        // number of arguments.
+        return inputB.call(
+                   std::vector<te::ExprHandle>({n, c, oh + r, ow + s})) *
+            filterB.call(std::vector<te::ExprHandle>({k, c, r, s}));
```
Nit: you should be able to use `filterB.load(k, c, r, s)` I think (same for `inputB`).

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

zheng-xq commented on test/cpp/tensorexpr/test_conv.cpp:
```diff
@@ -0,0 +1,86 @@
+  constexpr int S = 3;
+
+  // Output dims.
+  constexpr int OH = H - R + 1;
```
I just saw your response. This is perfectly fine. But I think the ones with padding to keep the output size the same is also very popular. Might be worth to have another one reflecting that. 

The reason a mask is interesting for performance in that when the axis is reordered, the masks should be adjusted as well.

bertmaher replied:
Oh I totally think it’s worth having some padded (and strided, grouped, etc) convolution tests too.  I agree :-)

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




Differential Revision: [D24142405](https://our.internmc.facebook.com/intern/diff/D24142405)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a 2D convolution example test"




Differential Revision: [D24142405](https://our.internmc.facebook.com/intern/diff/D24142405)

[ghstack-poisoned]
